### PR TITLE
eliminate uses of full from base/linalg/lq.jl

### PR DESCRIPTION
**Pull request by Sacha0**
Another small step towards deprecation of `full`. Ref. JuliaLang/LinearAlgebra.jl#229, JuliaLang/LinearAlgebra.jl#231, JuliaLang/julia#18850, and linked threads. Best!

### CODE REVIEW

File: base/linalg/lq.jl [hunk 1]
```diff
@@ -90,19 +88,25 @@ convert(::Type{LQPackedQ{T}}, Q::LQPackedQ) where {T} = LQPackedQ(convert(Abstra
...
+    methods with the implicit, packed representation of `Q`, may be less efficient and/or
+    require unnecessary allocations.
+
+!!! note "Implementation"
+    This method initially constructs a dense identity matrix, even though that may seem
+    inefficient, because every element in the returned matrix is a function of every element
+    in the (implicit) input matrix `Q`. That identity matrix is modified by the elementary
+    reflectors held in `Q`, so this operation does not reduce to simple indexing.
+"""
+squareQ(Q::LQPackedQ) = A_mul_B!(Q, eye(eltype(Q), reverse(size(Q.factors))...))
```
Comment (andreasnoack) [inline base/linalg/lq.jl]: This is not square. Should probably be `eye(eltype(Q), size(Q.factors, 2)).`

**Review by andreasnoack — CHANGES_REQUESTED** [review]
I'm also wondering if requesting the full matrix from `LQPackedQ` is a sufficiently common operation to require a separate function.

File: base/linalg/lq.jl (hunk 1 above)
Comment (Sacha0) [reply base/linalg/lq.jl]: Good catch! I failed to read the existing code critically. And agreed, `eye(eltype(Q), size(Q.factors, 2))` looks correct. What was the object the original code returns?

**Comment by Sacha0** [comment]
> I'm also wondering if requesting the full matrix from LQPackedQ is a sufficiently common operation to require a separate function.

To clarify, are you suggesting `squareQ(Q::LQPackedQ)` should not exist, or something beyond `squareQ(Q::LQPackedQ)` should exist? Thanks!

File: base/linalg/lq.jl (hunk 1 above)
Comment (andreasnoack) [reply base/linalg/lq.jl]: It looks like this PR is turning into a significant bug fix. The current version gives
```julia
julia> full(lqfact(randn(4,2))[:Q], thin = true)
2×2 Array{Float64,2}:
 -0.759173  0.650889
  0.650889  0.759173

julia> full(lqfact(randn(4,2))[:Q], thin = false)
2×4 Array{Float64,2}:
 -0.652378  -0.757894  0.0  0.0
 -0.757894   0.652378  0.0  0.0

julia> full(lqfact(randn(2,4))[:Q], thin = false)
4×2 Array{Float64,2}:
 -0.22708   -0.390319
  0.72372   -0.204318
 -0.025446   0.888878
 -0.651166  -0.125704

julia> full(lqfact(randn(2,4))[:Q], thin = false)
4×2 Array{Float64,2}:
 -0.0604746   0.111308
  0.822683   -0.406272
 -0.158067   -0.81894
  0.542725    0.389731
```
so only the first one is correct. Here is how they are supposed to look like

**Comment by andreasnoack** [comment]
> To clarify, are you suggesting `squareQ(Q::LQPackedQ)` should not exist

Yes. I'm suggesting that it shouldn't exist. I think it is pretty rare that you actually want to do that. If you, for instructional purposes, want to work with the full `Q` you would probably call `qr` instead of `qrfact` in the first place. We could also mention in the documentation that the dense `Q` can be computed from a package `Q` via multiplication with `eye`. It is very simple and also instructive since it would be how we'd compute `Q` anyway.

**Comment by Sacha0** [comment]
> Yes. I'm suggesting that it shouldn't exist. I think it is pretty rare that you actually want to do that. If you, for instructional purposes, want to work with the full Q you would probably call qr instead of qrfact in the first place. We could also mention in the documentation that the dense Q can be computed from a package Q via multiplication with eye. It is very simple and also instructive since it would be how we'd compute Q anyway.

Sounds good on this end :).

**Comment by Sacha0** [comment]
Revised. Thoughts? Thanks! :)

**Comment by kshyatt** [comment]
@andreasnoack as the author of that function, I included it for consistency with `qr.jl` but have no particular attachment to it otherwise.

**Review by andreasnoack — APPROVED** [review]

**Comment by Sacha0** [comment]
Thanks all! :)

**Pull request merged**